src/components/smart/Bluetooth.js

```diff
@@ -139,32 +187,73 @@ export default class Bluetooth extends Component {
      * Method that connects to a device with a given identifier (id)
      *
      * TODO: (4 pts)
-     *  (1) Connect to the device
-     *  (2) Retrieve services and characteristics
-     *  (3) Return data from (2)
-     *  (4) Correctly handle possible exceptions
+     *  (1) Connect to the device x
+     *  (2) Retrieve services and characteristics x
+     *  (3) Return data from (2) x
+     *  (4) Correctly handle possible exceptions x
      *
      * @param id
      * @returns {Promise<"react-native-ble-manager".PeripheralInfo>}
      */
     async connect(id){
-        throw new Error("Bluetooth: connect to device not implemented");
+      if (peripheral.connected){
+        BleManager.disconnect(peripheral.id);
+      }else{
+        BleManager.connect(peripheral.id).then(() => {
+          let peripherals = this.state.peripherals;
+          let p = peripherals.get(peripheral.id);
+          if (p) {
+            p.connected = true;
+            peripherals.set(peripheral.id, p);
+            this.setState({peripherals});
+            }
+            BleManager.retrieveServices(peripheral.id).then((peripheralInfo) => {
+              console.log(peripheralInfo);
+              const service = '13333333-3333-3333-3333-333333333337';
+              const bakeCharacteristic = '13333333-3333-3333-3333-333333330003';
+              const crustCharacteristic = '13333333-3333-3333-3333-333333330001';
+
+            setTimeout(() => {
+              BleManager.startNotification(peripheral.id, service, bakeCharacteristic).then(() => {
+                console.log('Started notification on ' + peripheral.id);
+                setTimeout(() => {
+                  BleManager.write(peripheral.id, service, crustCharacteristic, [0]).then(() => {
+                  BleManager.write(peripheral.id, service, bakeCharacteristic, [1,95]).then(() => {
+                    });
+                  });
+                }, 1000);
+              });
+            },  500);
```

Comment:
This looks like a major copy/paste... and was not even asked for. Also did not return characteristics. Didn't handle exceptions correctly.
1.5/4